src/auth.rs

```diff
@@ -99,28 +97,72 @@ impl Middleware<crate::MiniserveConfig> for Auth {
                     Err(err) => {
                         let auth_err = ContextualError::HTTPAuthenticationError(Box::new(err));
                         return Ok(Response::Done(
-                            HttpResponse::BadRequest().body(auth_err.to_string()),
+                            HttpResponse::BadRequest()
+                                .body(build_unauthorized_response(&req, auth_err, true, StatusCode::BAD_REQUEST)),
                         ));
                     }
                 };
                 if !match_auth(auth_req, required_auth) {
-                    let new_resp = HttpResponse::Unauthorized().finish();
-                    return Ok(Response::Done(new_resp));
+                    return Ok(Response::Done(HttpResponse::Unauthorized().body(
+                        build_unauthorized_response(
+                            &req,
+                            ContextualError::InvalidHTTPCredentials,
+                            true,
+                            StatusCode::UNAUTHORIZED,
+                        ),
+                    )));
                 }
             } else {
                 let new_resp = HttpResponse::Unauthorized()
                     .header(
                         header::WWW_AUTHENTICATE,
                         header::HeaderValue::from_static("Basic realm=\"miniserve\""),
                     )
-                    .finish();
+                    .body(build_unauthorized_response(
+                        &req,
+                        ContextualError::InvalidHTTPCredentials,
+                        false,
+                        StatusCode::UNAUTHORIZED,
+                    ));
                 return Ok(Response::Done(new_resp));
             }
         }
         Ok(Response::Done(resp))
     }
 }
 
+/// Builds the unauthorized response body
+/// The reason why log_error_chain is optional is to handle cases where the auth pop-up appears and when the user clicks Cancel.
+/// In those case, we do not log the error to the terminal since it does not really matter.
+fn build_unauthorized_response(
+    req: &HttpRequest<crate::MiniserveConfig>,
+    error: ContextualError,
+    log_error_chain: bool,
+    error_code: StatusCode,
+) -> String {
+    let error = ContextualError::HTTPAuthenticationError(Box::new(error));
+
+    if log_error_chain {
+        errors::log_error_chain(error.to_string());
+    }
+    let return_path = match &req.state().random_route {
+        Some(random_route) => format!("/{}", random_route),
+        None => "/".to_string(),
+    };
+
+    renderer::render_error(
+        &error.to_string(),
+        error_code,
+        &return_path,
+        None,
+        None,
+        req.state().default_color_scheme,
```

Comment:
I think this line is a bug. It should be the current theme, not default.

Reply:
Is there a use case where the user is not authenticated to the server, but as a custom theme set in the URL ? The only case I see would be to have an instance of miniserve running, set a custom theme, shutdown miniserve, and restart it, but that's kind of a corner case.

Note you'll observe the same problem on the page 404. I'll try to see how easy it is to parse the URL to retrieve query parameters, because at this stage of the code, that's the only way I can do it :)

Reply:
401 now does not have the back link as you suggested, and I fixed it for 404. Pushing asap.

Reply:
> Is there a use case where the user is not authenticated to the server, but as a custom theme set in the URL ? The only case I see would be to have an instance of miniserve running, set a custom theme, shutdown miniserve, and restart it, but that's kind of a corner case.
> 
> Note you'll observe the same problem on the page 404. I'll remove the back link on 401 as you suggested, and fix it for 404.

Fair point. I suppose the custom theme doesn't make much sense there.